model/test_results.go

```diff
@@ -579,6 +586,117 @@ func FindTestResults(ctx context.Context, env cedar.Environment, opts FindTestRe
 	return results, nil
 }
 
+// FindTestSamplesOptions specifies test results to collect test names from
+// and regex filters to apply to the test names.
+type FindTestSamplesOptions struct {
+	Tasks           []FindTestResultsOptions
+	TestNameRegexes []string
+}
+
+func (opts *FindTestSamplesOptions) createFindQuery() map[string]interface{} {
+	findQueries := []bson.M{}
+	for _, task := range opts.Tasks {
+		findQueries = append(findQueries, task.createFindQuery())
+	}
+
+	return bson.M{"$or": findQueries}
+}
+
+func (opts *FindTestSamplesOptions) createFindOptions() *options.FindOptions {
+	return options.Find().SetProjection(bson.M{
+		bsonutil.GetDottedKeyName(testResultsInfoKey, testResultsInfoTaskIDKey):        1,
+		bsonutil.GetDottedKeyName(testResultsInfoKey, testResultsInfoDisplayTaskIDKey): 1,
+		bsonutil.GetDottedKeyName(testResultsInfoKey, testResultsInfoExecutionKey):     1,
+		testResultsFailedTestsSampleKey:                                                1,
+	})
+}
+
+func (opts *FindTestSamplesOptions) makeTestSamples(testResults []TestResults) []TestResultsSample {
+	samples := consolidateSamples(testResults)
+	if len(opts.TestNameRegexes) > 0 {
+		samples = filterTestNames(samples, opts.TestNameRegexes)
+	}
+
+	return samples
+}
+
+func consolidateSamples(testResults []TestResults) []TestResultsSample {
+	type taskExecutionPair struct {
+		taskID    string
+		execution int
+	}
+	taskPairMap := make(map[taskExecutionPair][]string)
```

Comment:
Nit: we can clean this up a bit by creating the samples slice in the same for loop.
```
taskPairMap := make(map[taskExecutionPair]int) // where int is the index of the task/exec pair in the samples slice
samples := []TestResultsSample{}
for _, result := range testResults {
	pair := taskExecutionPair{
		taskID:    taskID,
		execution: result.Info.Execution,
	}
        idx, ok := taskPairMap[pair]
        if ok {
                samples[idx].FailedTestNames = append(samples[idx].FailedTestNames, results.FailedTestsSamples...)
        } else {
                samples = append(samples, TestResultsSample{
			TaskID:          pair.taskID,
			Execution:       pair.execution,
			FailedTestNames: names,
		})
                taskPairMap[pair] = len(samples) - 1
        }
}
```

Reply:
this is neat.
I refactored the `consolidateSamples` function (for the execution task comment above) and I'm not sure this comment is still applicable.